### ISSUE TITLE
SignalR .NET client does not transfer cookies between requests

### ISSUE THREAD

**Issue by cwt237** [SignalR/SignalR#3258]
It seems that the SignalR .NET client does not transfer cookies from previous requests to the next requests, e.g. a cookie returned in the negotiate response is not passed on for the following connect request (or any later reconnect/disconnect requests).

This is a problem when trying to achieve sticky cookie-based load balancing (the load balancer never sees the cookie, so it redirects the request to a random node behind it).

It works  fine with the JavaScript client, since the browser automatically passes along the cookies for that domain.

I am using SignalR 2.0.3 (but also checked code for SignalR 2.1.1 - it does not transfer the cookie as well).

The only work around (that I can think of) is to make an initial request to the load balancer - get the cookie and manually add the cookie to the CookieContainer on the connection object before making the negotiate request (then it should also use that cookie for the following connect, reconnect and disconnect), but that means that the client has to be aware of how the load balancing works.

**Comment by birdman7260** [comment]
I am also running into this issue. I'm using haproxy to set a cookie for load balancing. This obviously works fine for my browser clients but my .NET client is ignoring the cookie.

The workaround that @cwt237 described is what I've discovered I have to do, which means I will unfortunately have to force an update to my users so that I can load balance my service.

This seems like something that should be easily built into SignalR.

**Comment by BrennanConroy** [comment]
If you set the `CookieContainer` on the `HubConnection` before calling `HubConnection.Start()` then the cookies should flow through every request
e.g.
```
var hubConnection = new HubConnection("http://localhost:40476/");
var hubProxy = hubConnection.CreateHubProxy("TestHub");
hubConnection.CookieContainer = new System.Net.CookieContainer();
hubConnection.Start().Wait();
```

If this doesn't work for you then reopen please.

**Comment by birdman7260** [comment]
@BrennanConroy When I was investigating this I set the cookie fine using the CookieContainer. The issue, however, is that SignalR does not respect the Set-Cookie header from the response.

I would expect SignalR to, if the response to the GET on "negotiate" contained a Set-Cookie header, set that cookie for the rest of the process (e.g. the next GET on "start"). 

When I investigated whether that would work, setting the CookieContainer hadn't worked the way you show in your example. Are you certain that it should?

**Comment by BrennanConroy** [comment]
On the server I set a response cookie on `/negotiate`, after that `/start`, `/poll`, etc. all had the cookie as long as I set the `hubConnection.CookieContainer` before starting.

**Comment by kkaung** [comment]
I'm having trouble with this.  I'm using `Microsoft.AspNet.SignalR.Client 2.2.2` on .NET Framework 4.6.2 as my client app and `Microsoft.AspNet.SignalR 2.2.2` on ASP.NET MVC .NET Framework 4.6 as my server.  I don't see my cookie on the server side.

```
var uri = new Uri("http://someurl");
using (var connection = new HubConnection(hostUri.ToString()))
{
	var diagnosticProxy = connection.CreateHubProxy("DiagnosticHub");

	/**
	 * Start
	 */

	Cookie authCookie = GetAuthCookie();
	if (authCookie != null)
	{
		connection.CookieContainer = new CookieContainer();
		connection.CookieContainer.Add(authCookie);
	}

	connection.Start().Wait();

	Console.WriteLine("Started!");
	Console.ReadLine();
}
```

**Comment by NArnott** [comment]
I'm using 2.2.2 as well and trying to get it to work with my load using a sticky-session cookie.  I see my sticky session cookie getting stored in the CookieContainer but it's not getting passed backed to the server.

**Comment by kkaung** [comment]
I resolved my issue.  It was just a silly mistake.  Make sure the cookie path is pointed to the root of your domain or a certain path of your domain you are making it available to.

**Comment by sureshkumarbala** [comment]
@kaunglvlv  I am running into the same issue. We have the SignalR running fine in the test environment but breaks in the production environment. When we investigated the cookie returned empty only for the production environment with the same exact code.
Should I add the cookie manually for each request or something else going on in production infra?

**Comment by bdarby22** [comment]
I'm still using 2.2.2 as well.  Everything has been working beautifully for years until this week when I added authorization to a second service running on a different port.  My token authentication service is on 8081, and the resource services requiring authorization are also on 8081 and have never had a problem.  When I introduced the new services on 8082 the presence of the cookie in the /negotiate, /connect, and /poll calls are sometimes there and sometimes not there.  Everything still works as it should, but my logs are getting filled with errors when the cookie is missing.
